lib/common/acl.c

```diff
@@ -522,20 +534,22 @@ pcmk__post_process_acl(xmlNode *xml)
             crm_trace("ACLs allow creation of <%s> with id=\"%s\"",
                       crm_element_name(xml), display_id(xml));
 
-        } else {
+        } else if (check_top) {
             crm_trace("ACLs disallow creation of <%s> with id=\"%s\"",
                       crm_element_name(xml), display_id(xml));
-            if (xml != xmlDocGetRootElement(xml->doc)) {
-                pcmk_free_xml_subtree(xml);
-            }
+            pcmk_free_xml_subtree(xml);
             return;
+
+        } else {
+            crm_trace("ACLs would disallow creation of <%s> with id=\"%s\"",
+                      crm_element_name(xml), display_id(xml));
```

Comment:
2. I think we should bump the log priority for the last case.

Something should not have been present per ACL declarations to begin
with, and just in order to prevent explosions at the long-established
call sites, we cannot readily afford to block it (for destroying
possible assumptions), so we could at least be more verbose, for our
own affirmations about which assumptions are out there (with 100%
[asymmetric, on the safe side] reliability , vs. eye stepping through
the code with frequent cross-file jumps etc.) if for no other reason.

And since that message would be mostly for us, it could as well be
accompanied with `xml != xmlDocGetRootElement(xml->doc)` information
as inspected formerly for good measure (since that would also indicate
which of such achieved states could previously be the UAF issue, at
least when not reliably precluded with your `check_top`
conditionalizing).

(making me realize that `check_top` is only slightly bit less
ambiguous than `check_root`, luckily the comment clarifies that)